collections/nemo_asr/nemo_asr/audio_preprocessing.py

```diff
@@ -0,0 +1,470 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.****
+"""
+This file contains neural modules responsible for preprocessing audio data.
+"""
+__all__ = ['AudioPreprocessor',
+           'AudioToMFCCPreprocessor',
+           'AudioToMelSpectrogramPreprocessor',
+           'AudioToSpectrogramPreprocessor',
+           'MultiplyBatch',
+           'SpectrogramAugmentation']
+
+import math
+import torch
+import torchaudio
+try:
+    from apex import amp
+except AttributeError:
+    print("Unable to import APEX. Mixed precision and distributed training "
+          "will not work.")
+
+from nemo.backends.pytorch import NonTrainableNM
+from nemo.core import Optimization
+from nemo.core.neural_types import *
+from .parts.features import FilterbankFeatures
+from .parts.spectr_augment import SpecAugment, SpecCutout
+
+
+class AudioPreprocessor(NonTrainableNM):
+    """
+    A base class for Neural Modules that performs audio preprocessing,
+    transforming the wav files to features.
+    """
+    @staticmethod
+    def create_ports():
+        input_ports = {
+            "input_signal": NeuralType({0: AxisType(BatchTag),
+                                        1: AxisType(TimeTag)}),
+
+            "length": NeuralType({0: AxisType(BatchTag)}),
+        }
+
+        output_ports = {
+            "processed_signal": NeuralType({0: AxisType(BatchTag),
+                                            1: AxisType(SpectrogramSignalTag),
+                                            2: AxisType(ProcessedTimeTag)}),
+
+            "processed_length": NeuralType({0: AxisType(BatchTag)})
+        }
+        return input_ports, output_ports
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self.disable_casts = (self._opt_level == Optimization.mxprO1)
+
+        self.torch_windows = {
+            'hann': torch.hann_window,
+            'hamming': torch.hamming_window,
+            'blackman': torch.blackman_window,
+            'bartlett': torch.bartlett_window,
+            'none': None,
+        }
+
+    @torch.no_grad()
+    def forward(self, input_signal, length):
+        if self.disable_casts:
+            with amp.disable_casts():
+                processed_signal = self.get_features(
+                        input_signal.to(torch.float), length)
+        else:
+            processed_signal = self.get_features(input_signal, length)
+
+        processed_length = self.get_seq_len(length)
+        return processed_signal, processed_length
+
+    def get_features(self, input_signal, length):
+        # Called by forward(). Subclasses should implement this.
+        raise NotImplementedError
+
+    def get_seq_len(self, length):
```

Comment:
Now this is a good class to not make abstract because all three children essentially are call the same get_seq_len function.

Reply:
Let's make define this function for the base class as `return torch.ceil(seq_len / self.hop_length).to(dtype=torch.long)` and remove this function from all children.
Let's make the base class init require either n_window_stride or hop_length and set self.hop_length inside the init.

Reply:
Done.